tensorflow/contrib/gdr/gdr_memory_manager.cc

```diff
@@ -285,12 +290,12 @@ Status GdrMemoryManager::Init() {
   }
 
 #if GOOGLE_CUDA
+  VisitableAllocator::Visitor cuda_alloc_visitor =
+      std::bind(&GdrMemoryManager::InsertCUDAMemoryRegion, this, _1, _2);
```

Comment:
If I understand what you're doing correctly, you're registering the entire GPU memory to have the CU_POINTER_ATTRIBUTE_SYNC_MEMOPS property.  This seems to be a technique that NVIDIA provides because it makes concurrent programming easier, but I would avoid it because it likely will also damage performance considerably.

It sounds like you suspect a race condition on either the RDMA read, or some use of the final value.   For background, here's how we generally handle this issue.

TensorFlow tries hard to use as few synchronous CUDA calls as possible, and to introduce as few sync points as possible.   The techniques used rely on careful use of streams. The general discipline is that 4 streams are created for each GPU, by the [StreamGroupFactory](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/gpu/gpu_device.cc#L182).  All compute Ops (i.e. CUDA kernels) are launched on the single compute stream, and the other three are used for memcpys: one for H2D, one for D2H, and one for D2D copies.  In the normal case where an op relies only on the completion of prior ops in the same stream, it can be launched async without danger or any other temporal dependency. In the case where it relies on prior completion of an op on another stream, we need to introduce a sync dependency, which can be done in more than one way.  If we need to wait for an ordinary compute op prior to e.g. a D2H copy, we can just introduce a wait on the current compute stream prior to the copy, as [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/gpu/gpu_util.cc#L214).   This will cause the i/o stream to wait for completion of all ops _pending at the time of call_ to complete before launching the next op added to itself.  

In the case where we need to wait for a memcpy to terminate on one stream before launching a compute on another, the technique used is [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/gpu/gpu_util.cc#L168).  We don't want the compute stream to stall until the i/o is done, so we wait for the i/o to complete before queuing the compute op.  The wait is accomplished by a call the the EventMgr.

In our internal networking which uses GPUDirect, we use the EventMgr to ensure that we don't start a send before the op writing the buffer to be read has completed.  In the other direction, our RPC system provides a callback that executes only after the target buffer area has been written, and we use that callback to trigger subsequent ops that want to read that area.